---
title: 'Foushee'
layout: page
---

# Foushee v1.0.0
Bioinformatics pipeline for reference-free SNP analysis of Group-A *Streptococcus* (GAS) isolates

## Data workflow:
![Foushee pipeline](/staphb_toolkit/assets/workflows/foushee/Foushee_v1.0.png)

## Reference-free SNP analysis for GAS isolates of the same emm-type
Foushee uses [Trimmomatic](http://www.usadellab.org/cms/?page=trimmomatic) and [BBDuk](http://seqanswers.com/forums/showthread.php?t=42776) to perform read trimming and adapter/PhiX removal Monroe prior to using [Mash](http://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0997-x), a fast genome distance estimation algorithm, to predict the taxonomic identity of all isolates by querying input read data against a pre-skecthed RefSeq database. This method allows for accurate taxonomic identification up to the species level. The [emm typer tool from Public Health England (PHE)](https://github.com/phe-bioinformatics/emm-typing-tool) is then used to make emm-type predictions for each (GAS) isolate. The cleaned read data are also used to perform de novo assemblies using [Shovill](https://github.com/tseemann/shovill) and assembly metrics are collected using [Quast](https://github.com/ablab/quast). <br />
<br />
Taxonomic prediction results are then used to bin GAS assemblies by emm-type. For emm-type groupings made up of two or more isolates (e.g. 2 samples identified as emm-type 11) reference-free Single-Nucleotide Polymorphism (SNP) analysis will be performed on each emm-type group independently using [kSNP3](https://www.ncbi.nlm.nih.gov/pubmed/25913206).  kSNP3 results are used to generate a pairwise SNP distance matrix using [snp-dists](https://github.com/tseemann/snp-dists).

For emm-type groupings made up of three or more isolates (e.g. 3 samples identified as emm-type 89), output from snp-dists and ksnp3 are curated into a single pdf report using the [StaPH-B cluster-report-env](https://hub.docker.com/r/staphb/cluster-report-env).
---

## Quick Start:

````
$ staphb-wf foushee foushee.py <input_dir> -o <output_dir>
````

`<input_dir>` can be the path to an input directory containing paired-end fastq read data or path to a mounted BaseSpace project.
If an `<output_dir>` is not provided, results will be written to a `foushee_run_<date>` directory.


## Other Options
```
--profile: Custom Nextflow profile.
--config, -c, Path to a custom Nextflow configuration file
--get_config: Get a Nextflow configuration template for Tredegar
--resume: Resume a previous run
```
## Output:
Foushee will write the assembly metrics tsv file to the specified `<output_dir>`. All other output will be organized into four subdirectories:
- `<output_dir>/cluster_analysis`: Final pdf cluster analysis report; png renderings of the SNP heatmap and final parsimony tree within a `images` subdirectory; ksnp3 and snp-dists output (within the `snps` subdirectory)  to the specified Sorted BAM files after minimap2 mapping and ivar primer trimming
- `<output_dir>/emmtyper`: XML output files from [emm typer tool from Public Health England (PHE)](https://github.com/phe-bioinformatics/emm-typing-tool)
- `<output_dir>/shovill`: De novo genome assemblies generated by shovill
- `<output_dir>/logs`: Nextflow execution report (`Monroe_execution_report.html`), trace file (`Monroe_trace.txt`), and task work directories.


## Sample Assembly Metrics
![sample assembly metrics report](/staphb_toolkit/assets/workflows/foushee/assembly_metrics_sample.png)
- sample: isolate ID pulled from the fastq file
- est_genome_length: Average Q-score for the forward and reverse reads, respectively, calculated by CG Pipeline
- number_contigs: Total number of contigs in assembly calculated by Quast
- species_prediction: Genus and species prediction by Mash
- predicted_emmtype: emmtype prediction made by PHE's emm typer


## Docker Images
The base Nextflow configuration profiles (Docker, Singularity, and AWS) incorporate the following StaPH-B Docker Images:

| Tredegar Process   | Function  | Docker Image  | Comment|
|---|---|---|---|---|
| preProcess  | Renames input read files for downstream processing | staphb/fastqc:0.11.8  | Light-weight container for quick text processing  |
| mash_dist  | Calculates mash distances for each sketch against a pre-skecteched RefSeq database  | staphb/mash:2.1  | `mash dist` default parameters used to query against the pre-sketch RefSeq database available in the staphb/mash:2.1 Docker Image (/db) |
| mash_species  | Makes taxonomic species predictions based on closest mash distances | staphb/tiptoft:1.0.0  | Light-weight container with python3  |
| trim  | Quality trimming of input read data with bbduk  | staphb/trimmomatic:0.39  | `trimmomatic` parameters set to: minlength=100, windowsize=10, & qualitytrimscore=20|
| cleanreads  | Adapter and PhiX removal from input read data  | staphb/bbtools:38.76  |`bbduk` default parameters used |
| shovill | De novo genome assembly  | staphb/shovill:1.0.4  |`shovill.py` default parameters used |
| quast  | Quality assessment of de novo genome assemblies  | staphb/quast:5.0.2  |`quast` default parameters used |
| emmtype_finder  | Emm-type predictions for GAS isolates  | staphb/emm-typing-tool:0.0.1  | `emm_typing.py` default parameters used to query against the GAS database available in the staphb/emm-typing-tool:0.0.1 Docker Image (/db)|
| assembly_results | Curate assembly metrics and taxon predictions into a single report   |  staphb/tiptoft:1.0.0  | Light-weight container with python3  |
| ksnp3  | Performing reference-free SNP analysis  | staphb/mash:2.1'  | `ksnp3` parameters set to: kmer_length (-k)=19 parameters used, core SNPs only (--core)|
| snp-dists  | Generating pairwise snp-distance matrix from ksnp3 msa  | staphb/snp-dists:0.6.2  |`snp-dists` default parameters used |
| render  | Curating all output into a single pdf report | render_container = 'staphb/cluster-report-env:1.0'  | `SeqSero.py` default parameters used|


Default docker images and parameters listed above can be adjusted by:
1. Copying the template tredegar config file (`$ staphb-wf tredegar --get_config`)
2. Using a text editor to change the `<date>_tredegar.config` file
3. Specifying your custom config file (i.e. the edited `<date>_tredegar.config>` file) when running the pipeline:<br />

```
$ staphb-wf foushee <input_dir> -o <output_dir> -c <custom_config_file> [optoins]
```

## Version History

<b>Current version: v1.0.0 April 29, 2020</b>

Version 1.0.0 is the first stable version of Foushee

## Authors
[Kevin G. Libuit](https://github.com/kevinlibuit), DCLS Bioinformatics Lead Scientist <br />
