---
category: Workflows
path: '/Workflows/:id'
title: 'Tredegar'

layout: nil
---

# Tredegar v2.1
Bioinformatics pipeline for infectious disease WGS data QC

## Data workflow:
![Tredegar pipeline](assets/Tredegar_v2.1.png)

## Taxonomic Prediction:
Tredegar uses [Mash](http://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0997-x), a fast genome distance estimation algorithm, to predict the taxonomic identity of all isolates by querying input read data aginst the entire RefSeq databse. This method allows for accurate taxonomic identification up to the species level. 
For *Salmonella enterica*, *Escherichia coli*, and Group A *Streptoccus* isolates, subspecies characterization is determined using [SeqSero](http://jcm.asm.org/content/early/2015/03/05/JCM.00323-15), [SerotypeFinder](http://jcm.asm.org/content/53/8/2410.full.pdf+html), and [the emm typer tool from Public Health England (PHE)](https://github.com/phe-bioinformatics/emm-typing-tool) respectively. 

## Read Quality
De novo assemblies are performed using [Shovill](https://github.com/tseemann/shovill) and assembly metrics are collected using [Quast](https://github.com/ablab/quast). Quast calculations for assembly length are used to gauge genome coverage using the CDC's CG Pipeline--Q-score averages for the forward and reverse reads are also gauged using the [CG Pipeline](https://github.com/lskatz/CG-Pipeline).

---

## Quick Start: 

````
$ staphb_toolkit_workflows tredegar.py <input_dir> -o <output_dir>
````

`<input_dir>` can be the path to an input directory containing paired-end fastq read data or path to a mounted BaseSpace project.
If an `<output_dir>` is not provided, results will be written to a `tredegar_run_<date>` directory.


## Other Options
- -c: path to alternate configuration file, default set to staphb_toolkit/workflows/tredegar/tredegar_config.json
- -t: number of cpus to use, defaults to 8
- -m: number of GB of memory to use, defaults to 16

## Output: 
Tredegar will organize Mash, SeqyClean, CG Pipeline, Shovill, Quast, SeqSero, SerotypeFinder, and PHE's emm typer results in `<tool>_output` subdirectories within the `<output_dir>`, where `<tool>` is the name
of the bioinformatics tool used to generate the results. If more than one file is generated by a tool, a second subdirectory will be made within the `<tool>_output` directory for each sample. 
E.g. SerotypeFinder results for a Tredegar run on Sample001, and Sample002 will be organized like this:

`````
$tree serotypefinder_output

serotypefinder_output/
├── isolate001
│   ├── Hit_in_genome_seq.fsa
│   ├── results_table.txt
│   ├── results_tab.txt
│   ├── results.txt
│   └── Serotype_gene_seq.fsa
├── isolate002
│   ├── Hit_in_genome_seq.fsa
│   ├── results_table.txt
│   ├── results_tab.txt
│   ├── results.txt
│   └── Serotype_gene_seq.fsa

`````

A `mash_species.csv` file will also be generated and written within the `mash_output` directory. This file consists of the species predictions for each isolated made by MASH.

The final Tredegar Report will be written to a `tredegar_output` subdirectory within the `<output_dir>` along with a log file to assist in troubleshooting and output config file detailing run parameters.


## Sample Tredegar Report
![Sample_tredegar report](assets/tred2_sample_out.png)
- sample: isolate ID pulled from the fastq file
- r1_q and r2_q: Average Q-score for the forward and reverse reads, respectively, calculated by CG Pipeline
- est_genome_length: Shovill assembly length calculated by Quast
- est_cvg: Calculated genome coverage calculated by CG Pipeline
- number_contigs: Total number of contigs in assembly calculated by Quast
- predicted_species: Genus and species prediction by Mash
- predicted_serotype: Serotype predictions by SeqSero, SerotypeFinder, and PHE's emm typer for *Salmonella enterica* and *Escherichia coli* and Group A *Streptoccus* isolates, respectively. If a subspecies prediction cannot be made this field will be populated with "NA".  


---
## Version History

<b>Current version: v2.1 September 04, 2019</b>

Updates in v2.1
1. Written into the StaPH-B ToolKit
2. Report building using standard python libraries

Updates in [v2.0](https://github.com/kevinlibuit/Tredegar)
1. Use StaPH-B docker images rather than local downloads 
2. Incorporate CG-Pipeline for Q-score and coverage calculations
3. Incorporate Shovill and Quast to get accurate genome length estimations
4. Adjust E.coli serotyping to account for similar alleles
5. Use only O-type predictions from SerotypeFinder

Version 1.0 available [here](https://github.com/kevinlibuit/Tredegar/releases)

