---
title: 'Default Docker Images'
layout: page
---


| Application | Docker Image | Source |
|--------|-------| -------- |
| [ABRicate](https://hub.docker.com/r/staphb/abricate/)  | `staphb/abricate: 1.0.0 ` | https://github.com/tseemann/abricate |
| [Augur](https://github.com/nextstrain/augur)  | `staphb/auspice:6.3.0`| https://github.com/nextstrain/augur |
| [Auspice](https://github.com/nextstrain/auspice)  | `staphb/auspice:2.12.0` | https://github.com/nextstrain/auspice |
| [BBTools](https://hub.docker.com/r/staphb/bbtools/)  | `staphb/bbtools:38.76` | https://jgi.doe.gov/data-and-tools/bbtools/ |
| [BWA](https://hub.docker.com/r/staphb/bwa)  | `staphb/bwa:0.7.17` | https://github.com/lh3/bwa |
| [Canu-Racon](https://hub.docker.com/r/staphb/canu-racon/)  | `staphb/canu-racon:2.0` |https://github.com/isovic/racon  |
| [cfsan-snp-pipeline](https://hub.docker.com/r/staphb/cfsan-snp-pipeline)  | `staphb/cfsan-snp-pipeline:2.0.2`| https://github.com/CFSAN-Biostatistics/snp-pipeline |
| [Circlator](https://hub.docker.com/r/staphb/circlator)  | `staphb/circlator:1.5.6` | https://github.com/sanger-pathogens/circlator |
| [Clustalo](https://hub.docker.com/r/staphb/clustalo)  | `staphb/clustalo:1.2.4` | http://www.clustal.org/omega/ |
| [emm-typing-tool](https://hub.docker.com/r/staphb/emm-typing-tool)  | `staphb/emm-typing-tool:0.0.1` | https://github.com/phe-bioinformatics/emm-typing-tool |
| [FastANI](https://hub.docker.com/r/staphb/fastani)  |  `staphb/fastani:1.1`| https://github.com/ParBLiSS/FastANI |
| [FastTree](https://hub.docker.com/r/staphb/fasttree)  | `staphb/fasttree:2.1.11` | http://www.microbesonline.org/fasttree/ |
| [FastQC](https://hub.docker.com/r/staphb/fastqc)  |  `staphb/fastqc:0.11.8` | https://www.bioinformatics.babraham.ac.uk/projects/fastqc/ |
| [Filtlong](https://hub.docker.com/r/staphb/filtlong)  | `staphb/filtlong:0.2.0` | https://github.com/rrwick/filtlong |
| [Flye](https://hub.docker.com/r/staphb/flye)  |  `staphb/flye:2.7` | https://github.com/fenderglass/Flye |
| [iqtree](https://hub.docker.com/r/staphb/iqtree/)  |  `staphb/iqtree:1.6.7` | http://www.iqtree.org/ |
| [iVar](https://hub.docker.com/r/staphb/ivar/)  | `staphb/ivar:1.2.1` | https://github.com/andersen-lab/ivar |
| [iVar-SC2](https://hub.docker.com/r/staphb/ivar/)  | `staphb/ivar:1.2.1-SC2` | https://github.com/andersen-lab/ivar |
| [kma](https://hub.docker.com/r/staphb/kma/)  | `staphb/kma:1.2.21` | https://bitbucket.org/genomicepidemiology/kma/ |
| [Kraken](https://hub.docker.com/r/staphb/kraken/)  |  `staphb/kraken:1.0, 1.1.1` | https://github.com/DerrickWood/kraken |
| [Kraken2](https://hub.docker.com/r/staphb/kraken2/)  |  `staphb/kraken2:2.0.8-beta` | https://github.com/DerrickWood/kraken2 |
| [kSNP3](https://hub.docker.com/r/staphb/ksnp3/)  | `staphb/ksnp3:3.1` | https://sourceforge.net/projects/ksnp/ |
| [legsta](https://hub.docker.com/r/staphb/legsta/)  | `staphb/legsta:0.3.7` | https://github.com/tseemann/legsta |
| [Lyve-SET (includes CG-Pipeline scripts and raxml)](https://hub.docker.com/r/staphb/lyveset/)  | `staphb/lyvset:1.1.4f` | https://github.com/lskatz/lyve-SET https://github.com/lskatz/CG-Pipeline |
| [MAFFT](https://hub.docker.com/r/staphb/mafft/)  | `staphb/mafft:7.450` | https://mafft.cbrc.jp/alignment/software/ |
| [Mash](https://hub.docker.com/r/staphb/mash/)  |  `staphb/mash:2.2` | https://github.com/marbl/Mash |
| [mashtree](https://hub.docker.com/r/staphb/mashtree)  | `staphb/mashtree:1.0.4` | https://github.com/lskatz/mashtree |
| [medaka](https://hub.docker.com/r/staphb/medaka)  |  `staphb/medaka:0.8.1` | https://github.com/nanoporetech/medaka |
| [minimap2](https://hub.docker.com/r/staphb/minimap2)  |  `staphb/minimap2:2.17` | https://github.com/lh3/minimap2 |
| [mlst](https://hub.docker.com/r/staphb/mlst)  | `staphb/mlst:2.19.0` | https://github.com/tseemann/mlst |
| [Mugsy](https://hub.docker.com/r/staphb/mugsy)  | `staphb/mugsy:1r2.3` | http://mugsy.sourceforge.net/ |
| [MultiQC](https://hub.docker.com/r/staphb/multiqc)  | `staphb/multiqc:1.8` | https://github.com/ewels/MultiQC |
| [NanoPlot](https://hub.docker.com/r/staphb/nanoplot)  | `staphb/nanoplot:1.29.0` | https://github.com/wdecoster/NanoPlot |
| [NCBI AMRFinderPlus](https://hub.docker.com/r/staphb/ncbi-amrfinderplus)  | `staphb/ncbi-amrfinder-plus:3.1.1b` | https://github.com/ncbi/amr |
| [OrthoFinder](https://hub.docker.com/r/staphb/OrthoFinder)  |  `staphb/orthofinder:2.17` | https://github.com/davidemms/OrthoFinder |
| [Pilon](https://hub.docker.com/r/staphb/pilon)  | `staphb/pilon:1.23.0` | https://github.com/broadinstitute/pilon |
| [PlasmidSeeker](https://hub.docker.com/r/staphb/plasmidseeker)  | `staphb/plasmidseeker:1.0` | https://github.com/bioinfo-ut/PlasmidSeeker |
| [Prokka](https://hub.docker.com/r/staphb/prokka/)  |  `staphb/prokka:1.14.5` | https://github.com/tseemann/prokka |
| [QUAST](https://hub.docker.com/r/staphb/quast/)  |  `staphb/quast:5.0.2` | https://github.com/ablab/quast |
| [rasusa](https://hub.docker.com/r/staphb/rasusa/)  | `staphb/rasusa:0.2.0` | https://github.com/mbhall88/rasusa |
| [RAxML](https://hub.docker.com/r/staphb/raxml/)  | `staphb/raxml:8.2.12` | https://github.com/stamatak/standard-RAxML <br/> https://github.com/amkozlov/raxml-ng |
| [Roary](https://hub.docker.com/r/staphb/roary/)  |  `staphb/roary:3.12.0` | https://github.com/sanger-pathogens/Roary |
| [SalmID](https://hub.docker.com/r/staphb/salmid)  |  `staphb/salmid:0.1.23` | https://github.com/hcdenbakker/SalmID |
| [Samtools](https://hub.docker.com/r/staphb/samtools)  | `staphb/samtools:1.10` | https://github.com/samtools/samtools |
| [SeqSero](https://hub.docker.com/r/staphb/seqsero/)  |  `staphb/seqsero:1.0.1` | https://github.com/denglab/SeqSero |
| [SeqSero2](https://hub.docker.com/r/staphb/seqsero2/)  | `staphb/seqsero2:1.1.0` | https://github.com/denglab/SeqSero2/ |
| [seqyclean](https://hub.docker.com/r/staphb/seqyclean)  | `staphb/seqyclean:1.10.09` | https://github.com/ibest/seqyclean |
| [Seroba](https://hub.docker.com/r/staphb/seroba)  | `staphb/seroba:1.0.0` | https://github.com/sanger-pathogens/seroba |
| [SerotypeFinder](https://hub.docker.com/r/staphb/serotypefinder/)  | `staphb/serotypefinder:1.1` | https://bitbucket.org/genomicepidemiology/serotypefinder/ |
| [Shovill](https://hub.docker.com/r/staphb/shovill/)  | `staphb/shovill:1.1.0` | https://github.com/tseemann/shovill |
| [SISTR](https://hub.docker.com/r/staphb/sistr/)  | `staphb/sistr:1.0.2` | https://github.com/peterk87/sistr_cmd |
| [SKESA](https://hub.docker.com/r/staphb/skesa)  | `staphb/skesa:2.3.0` | https://github.com/ncbi/SKESA |
| [Snippy](https://hub.docker.com/r/staphb/snippy)  |  `staphb/snippy:4.6.0` | https://github.com/tseemann/snippy |
| [snp-dists](https://hub.docker.com/r/staphb/snp-dists)  | `staphb/snp-dists:0.6.2` | https://github.com/tseemann/snp-dists |
| [SNP-sites](https://hub.docker.com/r/staphb/snp-sites)  | `staphb/snp-sites:2.3.3` | https://github.com/sanger-pathogens/snp-sites |
| [SPAdes](https://hub.docker.com/r/staphb/spades/)  |  `staphb/spades:3.14.0` | http://cab.spbu.ru/software/spades/ |
| [SRA-toolkit](https://hub.docker.com/r/staphb/sratoolkit/)  | `staphb/sratoolkit:2.9.2` | https://github.com/ncbi/sra-tools |
| [Staramr](https://hub.docker.com/r/staphb/staramr/)  |  `staphb/staramr:0.7.1` | https://github.com/phac-nml/staramr |
| [TipToft](https://hub.docker.com/r/staphb/tiptoft/)  |  `staphb/tiptoft:1.0.0` | https://github.com/andrewjpage/tiptoft |
| [Trimmomatic](https://hub.docker.com/r/staphb/trimmomatic/)  | `staphb/trimmomatic:0.39` | http://www.usadellab.org/cms/?page=trimmomatic |
| [Unicycler](https://hub.docker.com/r/staphb/unicycler/)  | `staphb/unicycler:0.4.7` | https://github.com/rrwick/Unicycler |
| [wtdbg2](https://hub.docker.com/r/staphb/wtdbg2/)  |  `staphb/wtdbg2:2.5` | https://github.com/ruanjue/wtdbg2 |
