---
title: 'Default Docker Images'
layout: page
---



| Application   | Docker Image  | Tag  | Notes |
|---|---|---|---|---|
| Abricate | staphb/abricate | 0.8.13 | |
| BBTools |  |  |  |
| BWA |  |  |  |
| Canu-Racon |  |  |  |
| CFSAN-SNP |  |  |  |
| Circulator |  |  |  |
| ClustalO |  |  |  |
| Emm-typing-tool |  |  |  |
| FastANI |  |  |  |
| FastQC |  |  |  |
| FastTree |  |  |  |
| Filtlong |  |  |  |
| Flye |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |
|  |  |  |  |


'bbtools':'BBTools - Suite of fast, multithreaded bioinformatics tools for DNA and RNA sequence data',
'bwa':'BWA - mapping low-divergent sequences against a large reference genome',
'canu-racon':'Canu-Racon - Ultrafast consensus module for raw de novo assembly of long, uncorrected reads.',
'cfsan-snp':'CFSAN-SNP - SNP calling pipeline from the FDA CFSAN laboratory',
'circlator':'Circlator - A tool to circularize genome assemblies',
'clustalo':'ClustalO - A fast multiple sequence alignment program',
'emm-typing-tool':'Emm-typing-tool - Group A streptococci emm typing tool for NGS data',
'fastani':'FastANI - Fast whole-genome sequence average nucleotide identity (ANI) estimation',
'fastqc':'FastQC - A quality control tool for high throughput sequence data.',
'fasttree':'FastTree - Infers approximately-maximum-likelihood phylogenetic trees from alignments of nucleotide or protein sequences.',
'filtlong':'Filtlong - Quality filtering tool for long reads',
'flye':'Flye - De novo assembler for single molecule sequencing reads using repeat graphs',
'iqtree':'IQ-TREE - A fast and effective stochastic algorithm to infer phylogenetic trees by maximum likelihood.',
'kma':'KMA - Mapping method designed to map raw reads directly against redundant databases, in an ultra-fast manner using seed and extend.',
'ivar':'iVar - Computational package that contains functions broadly useful for viral amplicon-based sequencing.',
'ivar-SC2':'iVar - Computational package that contains functions broadly useful for viral amplicon-based sequencing: SARS-CoV-2 (SC2) reference sequence and ARTIC primers available in /reference ',
'kraken':'Kraken - Taxonomic sequence classification system ',
'kraken-build':'Kraken-build - Build a kraken database',
'kraken2':'Kraken2 - The second version of the Kraken taxonomic sequence classification system',
'kraken2-build':'Karken2-build - Build a kraken2 database',
'ksnp3':'kSNP2 - Identifies the pan-genome SNPs in a set of genome sequences, and estimates phylogenetic trees based upon those SNPs.',
'legsta':'Legsta - In silico Legionella pneumophila Sequence Based Typing',
'lyveset':'LYVE-SET - a method of using hqSNPs to create a phylogeny.',
'mash':'MASH - Fast genome and metagenome distance estimation using MinHash',
'mashtree':'MashTree - Create a tree using Mash distances',
'medaka':'Medaka - Sequence correction provided by ONT Research',
'minimap2':'Minimap2 - a versatile sequence alignment program that aligns DNA or mRNA sequences against a large reference database.',
'mlst':'MLST - Scan contig files against PubMLST typing schemes',
'mugsy':'Mugsy - A multiple whole genome aligner.',
'multiqc':'MultiQC - Aggregate results from bioinformatics analyses across many samples into a single report.',
'nanoplot':'NanoPlot - Plotting scripts for long read sequencing data ',
'ncbi-amrfinder-plus':'NCBI AMRFinderPlus - Designed to find acquired antimicrobial resistance genes and some point mutations in protein or assembled nucleotide sequences.',
'orthofinder':'OrthoFinder - Phylogenetic orthology inference for comparative genomics',
'pilon':'Pilon - Automated genome assembly improvement and variant detection tool',
'plasmidseeker':'PlasmidSeeker - A k-mer based program for the identification of known plasmids from whole-genome sequencing reads',
'prokka':'Prokka - Rapid prokaryotic genome annotation',
'quast':'Quast - Genome assembly evaluation tool.',
'rasusa':'RASUA - Randomly subsample sequencing reads to a specified coverage',
'raxml':'RAxML -Maximum likelihood tree builder.',
'roary':'Roary - Rapid large-scale prokaryote pan genome analysis.',
'salmid':'SalmID - Rapid confirmation of Salmonella spp. and subspp. from sequence data',
'samtools':'Samtools - A suite of programs for interacting with high-throughput sequencing data. It consists of three separate repositories.',
'seqsero':'SeqSero - Salmonella serotyping from genome sequencing data.',
'seqsero2':'SeqSero2 - Salmonella serotype prediction from genome sequencing data.',
'seqyclean':'SeqyClean - Pre-process and clean NGS data in order to prepare for downstream analysis',
'seroba':'Seroba - k-mer based Pipeline to identify the Serotype from Illumina NGS reads ',
'serotypefinder':'SerotypeFinder - identifies the serotype in total or partial sequenced isolates of E. coli.',
'shovill':'Shovill - Faster SPAdes assembler',
'sistr':'SISTR - Salmonella in silico typing resource command-line tool',
'skesa':'SKESA - NCBI\'s de novo genome assemlber',
'snippy':'Snippy - Rapid haploid variant calling and core genome alignment',
'snp-dists':'SNP-dists - Pairwise SNP distance matrix from a FASTA sequence alignment',
'snp-sites': 'SNP-sites - Finds SNP sites from a multi-FASTA alignment file',
'spades':'SPAdes - St. Petersburg genome assembler',
'sra-toolkit':'SRA ToolKit - Collection of tools and libraries for using data in the INSDC Sequence Read Archives.',
'staramr':'StarAMR - Scans genome contigs against the ResFinder, PlasmidFinder, and PointFinder databases.',
'tiptoft':'TipToft - Predict plasmids from uncorrected long read data',
'trimmomatic':'Trimmoamtic - Flexible read trimming tool for Illumina NGS data',
'unicycler':'Unicycler - an assembly pipeline for bacterial genomes.',
'wtdbg2':'WTDBG2 - Fuzzy Bruijn graph approach to long noisy reads assembly'
